### ISSUE TITLE
[naming-convention] Add support for `types` when array of selectors

### ISSUE THREAD

**Issue by mdziekon** [typescript-eslint/typescript-eslint#2377]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/naming-convention": [
      "error",
      {
        "selector": [ "variable", "property" ],
        "types": [ "boolean" ],
        "format": [ "PascalCase" ],
        "prefix": [ "is", "should", "has", "can", "did", "will" ]
      }
    ]
  }
}
```

```TS
const featureAEnabled = true;

class TestClass {
    featureAEnabled: boolean;
}
```

**Expected Result**

Using the specified rule configuration, ESLint is able to verify that my code (variable and class property) does not follow naming convention.

**Actual Result**

ESLint throws the following error:
```
Configuration for rule "@typescript-eslint/naming-convention" is invalid:
	Value {"selector":["variable","property"],"types":["boolean"],"format":["PascalCase"],"prefix":["is","should","has","can","did","will"]} should NOT have additional properties.
```

Looking at the schema code, it's clear that it in fact does not support `types` (nor `modifiers`), because this property of the configuration object is placed in a wrong nesting level:
https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/src/rules/naming-convention.ts#L255-L302

Both `modifiers` and `types` should be moved to begin at L281, not L283, at the same nesting level as `filter` and `selector`.
This looks like a mistake made in the original PR that introduced multi selectors (https://github.com/typescript-eslint/typescript-eslint/pull/2335/files#diff-6c2c7f346f804da11253beb86ccb0211R255), however... (see Additional Info).

**Additional Info**

At first this problem seems trivial to fix - I've already identified the offending schema lines that should be changed to make it work. However... both `types` and `modifiers` are a bit tricky specifiers in multi selector environment, since not every selector has the same `types` and `modifiers` available (eg. `function` selector has no `types`, while `variable` has few).

Therefore, changing the schema to what I've already proposed may solve the issue for rule configurations like mine, where every selector used has the same set of `types`, but it will also leave a validation hole for mixed selectors, leading to... undefined behaviour (probably). I can do that (as in, submit a PR with validation schema adjustment), but since it sounds controversial at best, I would like to hear other peoples' opinions first. I do not know if ESLint's validation schemas support dynamic validation (as it seems like the only logical solution that would satisfy both my needs and the need of configuration correctness; generating all possible permutations is most likely out of the question), but if it does, I'm also willing to work on this as long as someone can provide a bit of guidance on the topic.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.8.0` |
| `@typescript-eslint/parser`        | `3.8.0` |
| `TypeScript`                       | `3.9.3` |
| `ESLint`                           | `7.6.0` |
| `node`                             | `12.18.3` |
| `npm`                              | `6.14.6` |

**Comment by bradzacher** [comment]
#2376
